File: tests/r_parity/01_ols.py
"""StatsPAI OLS parity (Python side) -- Module 01.

Dumps sp.datasets.card_1995(simulated=True) to data/01_ols.csv and
runs sp.regress
with HC1 robust SEs. The companion 01_ols.R reads the same CSV and
runs lm() + sandwich::vcovHC(type="HC1"). Tolerance: rel < 1e-6
(closed-form estimator).
"""
from __future__ import annotations

import statspai as sp

from _common import ParityRecord, dump_csv, write_results


MODULE = "01_ols"
FORMULA = "lwage ~ educ + exper + expersq + black + south + smsa"


def main() -> None:
    # simulated=True is load-bearing, not decoration. This module *writes*
    # data/01_ols.csv, and the committed fixture -- together with the R and
    # Stata goldens computed on it -- is the calibrated replica. The
    # loader's default later became the real extract, so a bare
    # card_1995() call silently regenerates the fixture from different
    # data and breaks the same-byte premise of the row. Original-data
    # Card parity is a separate ledger (tests/orig_parity/01_card_original).
    df = sp.datasets.card_1995(simulated=True)
    dump_csv(df, MODULE)

    fit = sp.regress(FORMULA, data=df, robust="hc1")

    rows: list[ParityRecord] = []
    for name in fit.params.index:
        beta = float(fit.params[name])
        se = float(fit.std_errors[name])
        # Normalise the Python "Intercept" label to R's "(Intercept)"
        # so the parity comparator can join row-for-row.
        canonical = "(Intercept)" if name == "Intercept" else name
        rows.append(
            ParityRecord(
                module=MODULE,
                side="py",
                statistic=f"beta_{canonical}",
                estimate=beta,
                se=se,
                ci_lo=beta - 1.959963984540054 * se,
                ci_hi=beta + 1.959963984540054 * se,
                n=int(fit.data_info.get("n_obs", len(df))),
            )
        )

    write_results(MODULE, "py", rows, extra={"formula": FORMULA, "vcov": "HC1"})


if __name__ == "__main__":
    main()

File: tests/r_parity/01_ols.R
# StatsPAI OLS parity (R side) -- Module 01.
#
# Reads data/01_ols.csv (written by 01_ols.py) and runs lm() with
# sandwich::vcovHC(type = "HC1"). Tolerance: rel < 1e-6.

# Resolve _common.R relative to this script (works under Rscript and
# under interactive source() with chdir = TRUE).
.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) {
  dirname(normalizePath(sub("^--file=", "", .file_arg[1])))
} else {
  getwd()
}
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({
  library(sandwich)
})

MODULE  <- "01_ols"
FORMULA <- lwage ~ educ + exper + expersq + black + south + smsa

df <- read_csv_strict(MODULE)

fit <- lm(FORMULA, data = df)
vc  <- sandwich::vcovHC(fit, type = "HC1")
se  <- sqrt(diag(vc))

rows <- list()
for (name in names(coef(fit))) {
  beta <- unname(coef(fit)[name])
  s    <- unname(se[name])
  rows[[length(rows) + 1L]] <- parity_row(
    module    = MODULE,
    statistic = paste0("beta_", name),
    estimate  = beta,
    se        = s,
    ci_lo     = beta - qnorm(0.975) * s,
    ci_hi     = beta + qnorm(0.975) * s,
    n         = nrow(df)
  )
}

write_results(MODULE, rows, extra = list(formula = deparse(FORMULA), vcov = "HC1"))
